src/main/scala/loamstream/model/execute/RxExecuter.scala

```diff
@@ -54,58 +55,111 @@ final case class RxExecuter(
-    
-    //An Observable stream of job runs; each job is emitted when it becomes runnable.  This can be because the
-    //job's dependencies finished successfully, or because the job failed and we've decided to restart it.
-    //
-    //De-dupe jobs based on their ids and run counts.  This allows for re-running failed jobs, since while the
-    //job id would stay the same in that case, the run count would differ.  This is a blunt-force method that 
-    //prevents running the same job more than once concurrently in the face of "diamond-shaped" topologies.
-    val runnables: Observable[JobRun] = RxExecuter.deDupe(executable.multiplex(_.runnables))
+    def numJobs = executable.allJobs.size
     
-    //An observable stream of "chunks" of runnable jobs, with each chunk represented as a Seq.
-    //Jobs are buffered up until the amount of time indicated by 'windowLength' elapses, or 'runner.maxNumJobs'
-    //are collected.  When that happens, the buffered "chunk" of jobs is emitted.
-    val chunks: Observable[Seq[JobRun]] = runnables.tumblingBuffer(windowLength, runner.maxNumJobs, ioScheduler)
+    val executionState = TimeUtils.time(s"Initializing execution state with ${numJobs} jobs", debug(_)) {
+      ExecutionState.initialFor(executable, maxRunsPerJob)
+    }
+
+    val ioScheduler: Scheduler = IOScheduler()
     
-    val chunkResults: Observable[Map[LJob, Execution]] = for {
-      //NB: .toSet is important: jobs in a chunk should be distinct, 
-      //so they're not run more than once before transitioning to a terminal state.
-      jobs <- chunks.map(_.toSet)
-      //NB: Filter out jobs from this chunk that finished when run as part of another chunk, so we don't run them
-      //more times than necessary.  This helps in the face of job-restarting, since we can't call `distinct()` 
-      //on `runnables` and declare victory like we did before, since that would filter out restarting jobs that 
-      //already ran 
-      (finishedJobs, notFinishedJobs) = jobs.partition(_.status.isTerminal)
-      if notFinishedJobs.nonEmpty
-      (jobsToMaybeRun, skippedJobs) = notFinishedJobs.map(_.job).partition(jobFilter.shouldRun)
-      (jobsToCancel, jobsToRun) = jobsToMaybeRun.partition(jobCanceler.shouldCancel)
-      _ = handleSkippedJobs(skippedJobs)
-      cancelledJobsMap = cancelJobs(jobsToCancel)
-      _ = record(jobOracle, cancelledJobsMap)
-      skippedResultMap = toSkippedResultMap(skippedJobs)
-      executionTupleOpt <- runJobs(jobsToRun, jobOracle)
-      _ = record(jobOracle, executionTupleOpt)
-      executionMap = cancelledJobsMap ++ executionTupleOpt
-      _ = logFinishedJobs(executionMap)
-    } yield {
-      executionMap ++ skippedResultMap
+    val chunkResults: Observable[Map[LJob, Execution]] = {
+      //Note onBackpressureDrop(), in case runEligibleJobs takes too long (or the polling window is too short)
+      val ticks = Observable.interval(windowLength, ioScheduler).onBackpressureDrop
```

Comment:
Does `onBackpressureDrop` ever reset? Or can it get in a bad situation where you're waiting an arbitrarily long time due to one set of bad circumstances?

Reply:
Not that I know of, though there's probably a way to do that, either with RxScala or another Observable lib.  (Side note: RxScala is legacy at this point and will need to be replaced.)